### PR TITLE
refactor(final-project): SQL DDL and DML review

### PR DESCRIPTION
**Pull request by carson632**
We have a few minor issues, addressed in this branch.  Please feel free to go through and accept or decline changes, I won't `merge` until I get a yea or nay.

---------------------

But one major issue:  From Ethan's prior comment: 

> Reservation_Payment is therefore not necessary, since an instance of Payment is literally that. It's not a payment method, it's a single payment for a specific reservation.

Ethan wrote the DDL/DML to match this statement, while the ERD and RM still contain the AE `RESEVATION_PAYMENT`.

I think we're all agreed a `ReservationID` can have more than one `Payment` on it.  The business question we need to address is:

**_Can a single `payment` be applied to more than one `Reservation`?_**

If yes -> We need the AE and the script needs to change.
If no -> We need `ReservationID` in `PAYMENT` and the ERD and RM need to change.

-------------------------------------

Edit 20201201 19:19: Major issue resolved by this thread: https://docs.google.com/document/d/1F0xs6tqZ-q3HbtkfmdBNStHEKX1gr8TsuALYdAsWQfE/edit?disco=AAAAHXG95PQ

Minor things outstanding for reveiw.

Uploaded ERD too per this thread: https://docs.google.com/document/d/1F0xs6tqZ-q3HbtkfmdBNStHEKX1gr8TsuALYdAsWQfE/edit?disco=AAAAHXG95Pw

### CODE REVIEW

**Comment by ethandelany** [comment]
I agree with all of the changes except using VARCHAR for everything and tossing on LENGTH constraints. It's functionality is not equal to VARCHAR with drawbacks, it is more space efficient AND it enforces the length constraint anyways. Phone Number, State, and Zip Code should all just be NCHARs because they'll always be that length, never less. Otherwise, excellent job!

**Comment by carson632** [comment]
@ethandelany 

> it is more space efficient

I'll concede this point.  I think you're right here.  My experience is with SQL Server and Oracle, where the default behavior of `CHAR(X)` is to pad strings that are less than the length of X to be length X, thereby making them require X bytes, (or 2*X bytes if NCHAR), no matter what. Where as `VARCHAR(X)` will only require the length of the string in bytes, thereby making it more efficient in any case where `length(<text>) != X`. 
- _e.g._  'cat' -> 3 bytes in `VARCHAR(5)`, 5 bytes in `CHAR(5)` (and stored as 'cat  ').

In contrast, the default behavior of MySQL seems to trim whitespace for `CHAR(X)` fields upon entry. 
- _e.g._ 'cat  ' is and 5 bytes in `VARCHAR(5)`, 3 bytes in `CHAR(5)` (and stored as 'cat'),

------------------------------------------------

> it enforces the length constraint anyways.

This is not true.  Even in MySQL.

> Phone Number, State, and Zip Code should all just be NCHARs because they'll always be that length, never less.

This is the way it should be, but if you want to enforce it, you need the `CHECK CONSTRAINT`s.

Run the following:

```
CREATE TABLE IF NOT EXISTS CHAR_TEST (
	fieldId INT PRIMARY KEY AUTO_INCREMENT,
        fieldC CHAR(5),
        fieldV VARCHAR(5)
);

insert into CHAR_TEST
		values (1 ,'cat' ,'cat')
		      ,(2 ,'cat  ' ,'cat  ');
--------------------------------- This runs fine.  'cat' goes right into CHAR(5)
      
select * 
	from CHAR_TEST;
--------------------------------- see the data 

 select *
	from CHAR_TEST
        where length(fieldC) = 5;
--------------------------------- 0 records. For fieldC, 'cat' went in as is, but 'cat  ' was trimmed to 'cat'.


 select * 
	from CHAR_TEST
        where length(fieldC) = 3; 
 ---------------------------------2 records.  We entered 'cat' and 'cat  ' in fieldC, but the database removed the distinction.
```

-------------------------------------------------------------------

In conclusion, my concern with `CHAR(X)` is that 1) Though you'd think it should, it doesn't actually enforce the length to be X, and 2) It modifies your data without being told, albeit apparently in different ways depending on the RDBMS.  `VARCHAR()` always stores the string it's given, exactly as it's given.

For our project, this is a bit moot since you've generated "perfect" data.  If we actually had users entering their addresses/phones/whatever via a UI, I'm sure both you UI developers know how "perfect" that's going to come in haha.  I'm always going to advocate to avoid `char()` but feel free to `merge` from either your commit or mine.  Changing it either way really won't help or matter for this project.

**Comment by ethandelany** [comment]
Yeah, you are definitely right about the "not enforces length constraint" thing. I don't really know what I was thinking, but shortly after posting that I googled around and found myself to be incorrect. I kept in your CHECK constraints since that enforces minimum length and the size of CHAR enforces maximum length. 

VARCHAR does have some overhead to store the length of the String, so technically, in this scenario since we're always gonna have the exact length even if there's not perfect data, it is _slightly_ more efficient to use CHARs. It's like using an array vs a linked list.

It really doesn't matter honestly, and I'm very grateful that you're willing to compromise and work with me haha. I'll be accepting my pull request since I cleaned up a little, though I do admit your comments on what you changed and why were excellent and significantly helped me understand why you did what you did.

**Comment by carson632** [comment]
> the "not enforces length constraint" thing. I don't really know what I was thinking

You're thinking completely naturally.  If I told you, "_There's a field type established to _only_ have data 5 characters long entered into it._"  Would you expect it to accept the value 'cat'?  I don't think anybody's intuition is "_yes_".   In fact, if you want to know something dumb: I worked for over a year doing this **_professionally_** before I actually figured that out.  😂  

> VARCHAR does have some overhead to store the length of the String

I've heard of this too.  Specifically in the context of SQL Server, but I would have to assume it's at least similar in other RDBMS.    I've always assumed if I'm doubling space in some instances to go with `NVARCHAR` instead of `VARCHAR`, the one or two bytes shouldn't be a big deal.  I think "_it is slightly more efficient to use CHARs_" has to be right on principle.

-----------------------------------------

Perfect.  I think we're about done with what is required, yeah? After this we're all green on the doc.  Except for the queries we have to write against the database.
